tools/autograd/templates/python_torch_functions.cpp

```diff
@@ -534,4 +498,138 @@ void initTorchFunctions(PyObject* module) {
   }
 }
 
+/*
+ *
+ * Calls __torch_function__ on the overloaded arguments to a torch API
+ * function in order of precedence, returning the first result that is
+ * not NotImplemented. If all arguments return NotImplemented, raises a
+ * TypeError.
+ *
+ * Assumes overloaded_args has at least one entry. All entries must have
+ * a __torch_function__ attribute that resolves to a callable that
+ * accepts a torch API function, arguments, and keyword arguments for
+ * the torch API function.
+ *
+ * It is sufficient to call PythonArgs::has_torch_function before
+ * calling this function to verify that there are valid arguments
+ * present. If that is not done then special care must be taken to
+ * ensure there are arguments that are overloaded with
+ * __torch_function__.
+ *
+ * See torch._overrides._implement_torch_function for the equivalent
+ * code in the pure-python implementation.
+ *
+ * 'r' is a parsed PythonArgs instance, returned from
+ * PythonArgParser::parse.
+ *
+ * 'args' is a reference to the python tuple of arguments to the torch
+ * API function.
+ *
+ * 'kwargs' is a reference to the python dict of keyword arguments to
+ * the torch API function.
+ *
+ * 'torch_api' is a reference to python torch API namespace.
+ *
+ */
+
+PyObject* handle_torch_function(PythonArgs &r, PyObject* args, PyObject* kwargs, PyTypeObject &torch_api) {
+  py::object torch_api_function = PyObject_FastGetAttrString((PyObject*)&torch_api, const_cast<char*>(r.get_func_name().data()));
+  TORCH_INTERNAL_ASSERT(torch_api_function.ptr() != NULL, "torch API function must exist");
+  py::object ret = py::object();
+  for (auto &arg : r.overloaded_args) {
+    py::object torch_function = PyObject_FastGetAttrString(arg.ptr(), "__torch_function__");
+    ret = py::reinterpret_borrow<py::object>(PyObject_CallFunctionObjArgs(torch_function.ptr(), torch_api_function.ptr(), args, kwargs, NULL));
```

Comment:
I think `reinterpret_borrow` is the wrong cast function to use here. `reinterpret_borrow` means you don't take ownership of the returned pointer, whereas `reinterpret_steal` takes ownership. ` PyObject_CallFunctionObjArgs` returns a new reference. If you don't take ownership, you're going to leak it! Read the "py::handle (or subclass) → py::object (or subclass)" section from my cheat sheet https://github.com/pybind/pybind11/issues/1201; in fact, I think you can just say `ret = PyObject_CallFunctinObjArgs...` and the right thing will happen.

Reply:
Hmm, I think I had it as `reinterpret_steal` originally and I was seeing interpreter crashes, so I changed it to `reinterpret_borrow` which avoided the crash. `reinterpret_steal` will incref before taking ownership, but doesn't `PyObject_CallFunctionObjArgs` incref before returning the `PyObject*` instance? So we want to *not* incref when we create the `py::object`, and then decref when returning the result to the interpreter. Is my reasoning wrong here?

I haven't tried `ret = PyObject_CallFunctinObjArgs...`, will try that and report back...

Reply:
Yeah, I think the simple binding only works for `py::handle`, not `py::object`:

```
../torch/csrc/autograd/generated/python_torch_functions.cpp:1481:106: error: no match for ‘operator=’ (operand types are ‘pybind11::object’ and ‘PyObject* {aka _object*}’)
     ret = PyObject_CallFunctionObjArgs(torch_function.ptr(), torch_api_function.ptr(), args, kwargs, NULL);
```

I also just double-checked and I'm getting segmentation faults when I run `test_overrides` if I use `reinterpret_steal`.

Reply:
> `reinterpret_steal` will incref before taking ownership, but doesn't `PyObject_CallFunctionObjArgs` incref before returning the `PyObject*` instance?

Oops, this is backwards. I guess I need to more fully understand why we're getting seg faults when I use `reinterpret_steal`...

Reply:
I think I have it figured out. Since `handle_torch_function` returns a `PyObject*`, the `py::object` wrapper for the return value will get deallocated at the end of `handle_torch_function`, so there will be a decref at the end of this function no matter what. One way to handle this is to use `reinterpret_borrow` as I originally had it, that way we incref when we get the object and then decref when we return it, leaving the reference count at 1.

Another way to do it that's probably easier to understand is to explicitly incref when we return the result as a `PyObject*` but still use `reinterpret_steal`. I'll push an update shortly.

Reply:
Yes, the latter is semantically correct. Don't "leak a reference", and then reclaim the leak later ;) (As a clarifying note, once I saw the leak, I stopped reading the rest of the code haha)